src/vos/vos_obj_cache.c

```diff
@@ -686,3 +693,251 @@ vos_obj_evict_by_oid(struct vos_container *cont, daos_unit_oid_t oid)
+		D_ALLOC_ARRAY(ranges, bkts->vba_cnt);
+		if (ranges == NULL)
+			return -DER_NOMEM;
+	} else {
+		ranges = &rg_inline[0];
+	}
+
+	for (i = 0; i < bkts->vba_cnt; i++) {
+		D_ASSERT(bkts->vba_bkts[i] != UMEM_DEFAULT_MBKT_ID);
+		ranges[i].cr_off = umem_get_mb_base_offset(vos_pool2umm(pool), bkts->vba_bkts[i]);
+		ranges[i].cr_size = vos_pool2store(pool)->cache->ca_page_sz;
+	}
+
+	rc = umem_cache_pin(vos_pool2store(pool), ranges, bkts->vba_cnt, false, pin_hdl);
+	if (rc)
+		DL_ERROR(rc, "Failed to pin %u ranges.", bkts->vba_cnt);
+
+	if (ranges != &rg_inline[0])
+		D_FREE(ranges);
+
+	return rc;
+}
+
+static int
+vos_obj_acquire(struct vos_container *cont, daos_unit_oid_t oid, struct vos_object **obj_p)
+{
+	struct vos_object	*obj;
+	struct daos_lru_cache	*occ;
+	int			 rc;
+
+	D_ASSERT(cont != NULL);
+	D_ASSERT(cont->vc_pool);
+	D_ASSERT(obj_p != NULL);
+	*obj_p = NULL;
+
+	occ = vos_obj_cache_get(cont->vc_pool->vp_sysdb);
+	D_ASSERT(occ != NULL);
+
+	/* Lookup object cache, create cache entry if not found */
+	rc = obj_get(occ, cont, oid, true, &obj);
+	if (rc) {
+		DL_ERROR(rc, "Failed to lookup/create object in cache.");
+		return rc;
+	}
+
+	if (obj->obj_zombie) {
+		D_ERROR("The object:"DF_UOID" is already evicted.\n", DP_UOID(oid));
+		obj_put(occ, obj, true);
+		return -DER_AGAIN;
+	}
+
+	/* Lookup OI table if the cached object is negative */
+	if (obj->obj_df == NULL) {
+		obj->obj_sync_epoch = 0;
+		rc = vos_oi_find(cont, oid, &obj->obj_df, NULL);
+		if (rc == 0) {
+			obj->obj_sync_epoch = obj->obj_df->vo_sync;
+		} else if (rc == -DER_NONEXIST) {
+			rc = 0;
+		} else if (rc) {
+			DL_ERROR(rc, "Failed to lookup OI table.");
+			obj_put(occ, obj, false);
+			return rc;
+		}
+	}
+
+	obj_allot_bkt(cont->vc_pool, obj);
+	*obj_p = obj;
+
+	return 0;
+}
+
+struct vos_pin_handle {
+	unsigned int		 vph_acquired;
+	struct umem_pin_handle	*vph_pin_hdl;
+	struct vos_object	*vph_objs[0];
+};
+
+void
+vos_unpin_objects(daos_handle_t coh, struct vos_pin_handle *hdl)
+{
+	struct vos_container	*cont = vos_hdl2cont(coh);
+	struct vos_pool		*pool = vos_cont2pool(cont);
+	int			 i;
+
+	if (hdl->vph_pin_hdl != NULL)
+		umem_cache_unpin(vos_pool2store(pool), hdl->vph_pin_hdl);
+
+	for (i = 0; i < hdl->vph_acquired; i++)
+		vos_obj_release(hdl->vph_objs[i], 0, false);
+
+	D_FREE(hdl);
+}
+
+int
+vos_pin_objects(daos_handle_t coh, daos_unit_oid_t oids[], int count, struct vos_pin_handle **hdl)
+{
+	struct vos_pin_handle	*vos_hdl;
+	struct vos_object	*obj;
+	struct vos_bkt_array	 bkts;
+	struct vos_container	*cont = vos_hdl2cont(coh);
+	struct vos_pool		*pool = vos_cont2pool(cont);
+	int			 i, rc;
+
+	*hdl = NULL;
+	if (!vos_pool_is_evictable(pool))
+		return 0;
+
+	D_ASSERT(count > 0);
+	D_ALLOC(vos_hdl, sizeof(*vos_hdl) + sizeof(struct vos_object *) * count);
+	if (vos_hdl == NULL)
+		return -DER_NOMEM;
+
+	vos_bkt_array_init(&bkts);
+	for (i = 0; i < count; i++) {
+		rc = vos_obj_acquire(cont, oids[i], &vos_hdl->vph_objs[i]);
```

Comment:
Why not vos_obj_hold()? that will also pin the object, right?

Reply:
We don't want to pin object here.